### PR TITLE
Simplify segment creation and inheritance in dialects

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made

This is a big PR, but most of the changes are very mechanical.
- [X] Removes the `@segment()` decorator from the `Dialect` class, replacing it with a more automated approach.
- [X] Segment inheritance now uses normal Python syntax. Benefit: `mypy` understands this syntax, eliminating the need for a number of `# type: ignore` comments.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 1]
```diff
@@ -1001,15 +1003,9 @@ class IntoClauseSegment(BaseSegment):
 
 
 @postgres_dialect.segment(replace=True)
-class UnorderedSelectStatementSegment(BaseSegment):
+class UnorderedSelectStatementSegment(ansi.UnorderedSelectStatementSegment):
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_postgres.py:1006]: By using actual Python class inheritance, we eliminate a lot of duplicate code from the base class.

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 2]
```diff
@@ -1136,9 +1121,7 @@ class CreateRoleStatementSegment(BaseSegment):
 
 
 @postgres_dialect.segment(replace=True)
-class ExplainStatementSegment(
-    ansi_dialect.get_segment("ExplainStatementSegment")  # type: ignore
-):
+class ExplainStatementSegment(ansi.ExplainStatementSegment):
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_postgres.py:1124]: Here, we were already using real inheritance, but using a function to look up the base class. This caused `mypy` to complain, requiring a `# type: ignore` comment. Now we can remove that.

File: src/sqlfluff/dialects/dialect_postgres.py [hunk 3]
```diff
@@ -1001,15 +1003,9 @@ class IntoClauseSegment(BaseSegment):
...
 @postgres_dialect.segment(replace=True)
-class UnorderedSelectStatementSegment(BaseSegment):
+class UnorderedSelectStatementSegment(ansi.UnorderedSelectStatementSegment):
     """Overrides ANSI Statement, to allow for SELECT INTO statements."""
 
-    type = "select_statement"
-
-    match_grammar = ansi_dialect.get_segment(
-        "UnorderedSelectStatementSegment"
-    ).match_grammar.copy()
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_postgres.py:1011]: We don't need to _copy_ it because we now _inherit_ it.

**Comment by barrywhart** [comment]
@alanmcruickshank: Could you take a look at this approach and see what you think? It really cleans up the segment inheritance:
* Shorter code
* Makes the `mypy` type checker happy

IIRC, at some point we had concerns about circular dependencies with the dialect files, hence the creation of the `load_raw_dialect()` function. Do you recall the details? Does this PR potentially re-introduce the same issues?

**Comment by barrywhart** [comment]
@tunetheweb: After our discussion on the last PR, I started this experiment. I asked Alan for his feedback, since it touches on something that had caused us issues a year or more ago.

**Comment by tunetheweb** [comment]
I was thinking about this some more and wondering could we do more with the `replace=True` option? Could there be an `inherit=True` option as well? Or should replace=True do this by default?

**Comment by barrywhart** [comment]
About `replace` and `inherit`, I wonder if we could possibly eliminate the need for the `@segment` decorator entirely. A dialect is basically the module contents. Could we just define the module and create the dialect automatically? Overall, if we can make more use of built-in Python language features like modules and classes, that can make it easier to work on dialects (for example, mypy and Python IDEs will help with writing and checking the code).

**Review by alanmcruickshank — APPROVED** [review]
Nice. Love it 👍 

Good work.

File: src/sqlfluff/core/dialects/base.py [hunk 4]
```diff
@@ -196,6 +154,19 @@ def replace(self, **kwargs: DialectElementType):
                 raise ValueError(f"{n!r} is not already registered in {self!r}")
             self._library[n] = kwargs[n]
 
+    def add_update_segments(self, module_dct):
+        """Scans module dictionary, adding or replacing segment definitions.
+
+        Each dialect module (.py file) should call this function at the END OF
+        THE FILE, after all the segment classes have been defined.
```
Comment (tunetheweb) [inline src/sqlfluff/core/dialects/base.py:161]: Anyway we could automate this so we don't have to add? e.g. by adding to `load_raw_dialect`?

**Review by tunetheweb — APPROVED** [review]
One question, but in general LGTM.

This is a fairly big (but good!) change to dialects (even if the code change itself is small). So flagging to @sqlfluff/sqlfluff-maintainers (especially @alanmcruickshank).

File: src/sqlfluff/core/dialects/base.py (hunk 4 above)
Comment (barrywhart) [reply src/sqlfluff/core/dialects/base.py:161]: Good call -- calling it from `load_raw_dialect()` seems to be working fine. I'll update.

**Pull request merged**